src/server/http/xsrf.js

```diff
@@ -4,18 +4,48 @@ export default function (kbnServer, server, config) {
   const disabled = config.get('server.xsrf.disableProtection');
   const versionHeader = 'kbn-version';
   const xsrfHeader = 'kbn-xsrf';
+  const contentTypeHeader = 'content-type';
+  const allowedRequestMediaTypes = ['application/json', 'application/x-ndjson'];
 
   server.ext('onPostAuth', function (req, reply) {
     if (disabled) {
       return reply.continue();
     }
 
     const isSafeMethod = req.method === 'get' || req.method === 'head';
-    const hasVersionHeader = versionHeader in req.headers;
-    const hasXsrfHeader = xsrfHeader in req.headers;
+    if (!isSafeMethod) {
+      const hasVersionHeader = versionHeader in req.headers;
+      const hasXsrfHeader = xsrfHeader in req.headers;
 
-    if (!isSafeMethod && !hasVersionHeader && !hasXsrfHeader) {
-      return reply(badRequest(`Request must contain an ${xsrfHeader} header`));
+      if (hasXsrfHeader) {
+        server.log(
+          ['warning', 'deprecation'],
+          `The ${xsrfHeader} header is deprecated and will be removed in a future version of Kibana.` +
+          ` Specify a ${contentTypeHeader} header of either application/json or application/x-ndjson instead.`
```

Comment:
Two things:

1. I think we should explore testing the `server.log`s. We can maybe `sinon.stub` the `server.log` method and ensure it's called correctly. Would be great to check the `callCount` on the stub too.

2. When this is logged you _can_ already have a correct content-type header specified, so we shouldn't include the second line, as it might get people confused (hm, ideally as this is related to security we should link people to the docs where our handling of xsrf is explain? cc @epixa).

   For now, maybe add tests that cover:
   - `xsrfHeader` but not `contentTypeHeader`
   - `xsrfHeader` but disallowed `contentTypeHeader`
   - `xsrfHeader` and allowed `contentTypeHeader`